Applications/Spire/Source/Ui/OpenFilterPanel.cpp

```diff
@@ -0,0 +1,358 @@
+    std::bind_front(&OpenFilterPanel::on_mode_current, this));
+  mode_layout->addStretch();
+  layout->addLayout(mode_layout);
+  layout->addSpacing(scale_height(18));
+  m_input_box = input_box_builder(m_matches);
+  layout->addWidget(m_input_box);
+  layout->addSpacing(scale_height(8));
+  m_list_view = new ListView(m_matches,
+    std::bind_front(&OpenFilterPanel::make_item, this));
+  update_style(*m_list_view, [] (auto& style) {
+    style = LIST_VIEW_STYLE(style);
+  });
+  m_matches_connection = m_matches->connect_operation_signal(
+    std::bind_front(&OpenFilterPanel::on_matches_operation, this));
+  auto scrollable_list_box = new ScrollableListBox(*m_list_view);
+  scrollable_list_box->setMinimumSize(scale_width(164), scale_height(80));
+  scrollable_list_box->setMaximumHeight(scale_height(158));
+  layout->addWidget(scrollable_list_box);
+  m_filter_panel = new FilterPanel(std::move(title), this, parent);
+  m_filter_panel->connect_reset_signal(
+    std::bind_front(&OpenFilterPanel::on_reset, this));
+  auto panel_window = window();
+  panel_window->setWindowFlags(Qt::Tool | panel_window->windowFlags());
+  panel_window->installEventFilter(this);
+}
+
+const std::shared_ptr<AnyListModel>& OpenFilterPanel::get_matches() const {
+  return m_matches;
+}
+
+const std::shared_ptr<ValueModel<OpenFilterPanel::Mode>>&
+    OpenFilterPanel::get_mode() const {
+  return m_mode;
+}
+
+connection OpenFilterPanel::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+bool OpenFilterPanel::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::Close) {
+    m_filter_panel->hide();
+    hide();
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+bool OpenFilterPanel::event(QEvent* event) {
+  if(event->type() == QEvent::ShowToParent) {
+    m_filter_panel->show();
+    window()->activateWindow();
+    m_input_box->setFocus();
+  } else if(event->type() == QEvent::HideToParent) {
+    m_filter_panel->hide();
+  }
+  return QWidget::event(event);
+}
+
+QWidget* OpenFilterPanel::make_item(const std::shared_ptr<AnyListModel>& model,
+    int index) {
+  auto label = displayTextAny(model->get(index));
+  auto item = new DeletableItem(label);
+  item->connect_delete_signal([=] {
+    auto index = [&] {
+      for(auto i = 0; i < m_matches->get_size(); ++i) {
+        if(label == displayTextAny(model->get(i))) {
+          return i;
+        }
+      }
+      return -1;
+    }();
+    if(index >= 0) {
+      m_matches->remove(index);
+      m_input_box->setFocus();
+    }
+  });
+  return item;
+}
+
+void OpenFilterPanel::on_mode_current(Mode mode) {
+  submit();
+}
+
+void OpenFilterPanel::on_matches_operation(
+    const AnyListModel::Operation& operation) {
+  visit(operation,
+    [&] (const AnyListModel::AddOperation& operation) {
+      update_style(*m_list_view, [] (auto&) {});
```

Comment:
What does this do?

Reply:
When a ListItem is added to the ListView, the newly added ListItem cannot be styled, so I have to style the ListItem or update the style of the ListView when the add operation happens.

Reply:
That's good, when you come across bugs in the future, instead of trying to work around them, report them in Asana so they can get fixed.

Reply:
For this one, should I remove it or keep it?

Reply:
You can keep it, but I need to fix this bug, so in the future just find a way to make it clear that this is working around a bug, preferably by creating a task in Asana. In your Asana task you can refer to this line (along with the commit) and point out that this is a workaround to a bug.